### PR TITLE
Added protocol - JRMI - Java Remote Method Invocation

### PR DESCRIPTION
**Pull request by dandare100**
Please sign (check) the below before submitting the Pull Request:

- [X] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [X] I have read the contributing guide lines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [X] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Describe changes:

Added JRMI protocol.

https://docs.oracle.com/en/java/javase/21/docs/specs/rmi/protocol.html

### CODE REVIEW

**Comment by IvanNardi** [comment]
Not sure why CI tests are not triggered... It happened also with your previous PR...
Try to close and re-open...

**Comment by IvanNardi** [comment]
> Not sure why CI tests are not triggered... It happened also with your previous PR... Try to close and re-open...

@dandare100, are you pushing via API, using `GITHUB_TOKEN`?

File: src/include/ndpi_protocol_ids.h [hunk 1]
```diff
@@ -444,6 +444,7 @@ typedef enum {
   NDPI_PROTOCOL_THREADS               = 413,
   NDPI_PROTOCOL_VIBER_VOIP            = 414,
   NDPI_PROTOCOL_ZUG                   = 415,
+  NDPI_PROTOCOL_JRMI                  = 416, /*https://docs.oracle.com/en/java/javase/21/docs/specs/rmi/protocol.html*/
```
Comment (IvanNardi) [inline src/include/ndpi_protocol_ids.h]: Remove the comment, please. You already added it in the documentation

File: src/lib/protocols/jrmi.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*
+ * jrmi.c
+ *
+ * Copyright (C) 2011-24 - ntop.org
```
Comment (IvanNardi) [inline src/lib/protocols/jrmi.c]: Copyright (C) 2024 - ntop.org

File: src/lib/protocols/jrmi.c [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+#define NDPI_CURRENT_PROTO NDPI_PROTOCOL_JRMI
+
+#include "ndpi_api.h"
+#include "ndpi_private.h"
+
+
+static void ndpi_int_jrmi_add_connection(struct ndpi_detection_module_struct *ndpi_struct,
+					 struct ndpi_flow_struct *flow/* , */
+					 /* ndpi_protocol_type_t protocol_type */) {
```
Comment (IvanNardi) [inline src/lib/protocols/jrmi.c]: I know that you likely "copied" this code from somewhere else, but could you remove the third parameter, please?

File: src/lib/protocols/jrmi.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_JRMI, NDPI_PROTOCOL_UNKNOWN, NDPI_CONFIDENCE_DPI);
+}
+
+static void ndpi_search_jrmi(struct ndpi_detection_module_struct *ndpi_struct, struct ndpi_flow_struct *flow) {
+			     struct ndpi_packet_struct *packet = &ndpi_struct->packet;
+
+	NDPI_LOG_DBG(ndpi_struct, "search jrmi\n");
+
+	if (packet->tcp != NULL) {
```
Comment (IvanNardi) [inline src/lib/protocols/jrmi.c]: You can remove this check. We are sure that only TCP packets trigger this code

File: src/lib/protocols/jrmi.c [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+
+static void ndpi_int_jrmi_add_connection(struct ndpi_detection_module_struct *ndpi_struct,
+					 struct ndpi_flow_struct *flow/* , */
+					 /* ndpi_protocol_type_t protocol_type */) {
+	ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_JRMI, NDPI_PROTOCOL_UNKNOWN, NDPI_CONFIDENCE_DPI);
+}
+
+static void ndpi_search_jrmi(struct ndpi_detection_module_struct *ndpi_struct, struct ndpi_flow_struct *flow) {
+			     struct ndpi_packet_struct *packet = &ndpi_struct->packet;
```
Comment (IvanNardi) [inline src/lib/protocols/jrmi.c]: Indentation

File: src/lib/protocols/jrmi.c [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+				packet->payload[2] == 0x4d &&
+				packet->payload[3] == 0x49 ) {
+				NDPI_LOG_INFO(ndpi_struct, "found jrmi over tcp\n");
+				ndpi_int_jrmi_add_connection(ndpi_struct, flow);
+				return;
+			}
+		}
+	}
+
+	if(flow->packet_counter > 5)
```
Comment (IvanNardi) [inline src/lib/protocols/jrmi.c]: You can remove the check and always exclude the protocol here
AFSIU, the magic word + version fields are present only in the first packet of the flow. If we miss it, we will never find them again

**Review by IvanNardi — CHANGES_REQUESTED** [review]
As reported by the CI:
* you need to update the tests results
* could you fix the window build? You need to add the new c file in `windows/nDPI.vcxproj`

Some other minor issues below

**Comment by dandare100** [comment]
> > Not sure why CI tests are not triggered... It happened also with your previous PR... Try to close and re-open...
> 
> @dandare100, are you pushing via API, using `GITHUB_TOKEN`?

Hello there @IvanNardi 

I push to my nDPI fork (using an ssh key) and create the pull request on the github interface.
Should I change it to another way ?

Jeff

**Comment by IvanNardi** [comment]
> > > Not sure why CI tests are not triggered... It happened also with your previous PR... Try to close and re-open...
> > 
> > 
> > @dandare100, are you pushing via API, using `GITHUB_TOKEN`?
> 
> Hello there @IvanNardi
> 
> I push to my nDPI fork (using an ssh key) and create the pull request on the github interface. Should I change it to another way ?
> 
> Jeff

No you don't; it seems right. I was trying to understand why your commit/push/close-open doesn't trigger the GH Actions...

**Comment by dandare100** [comment]
PS : Apologies for the up and down on the commits. I missed the MS Visual studio stuff

File: src/include/ndpi_protocol_ids.h (hunk 1 above)
Comment (dandare100) [reply src/include/ndpi_protocol_ids.h]: Will do

File: src/lib/protocols/jrmi.c (hunk 2 above)
Comment (dandare100) [reply src/lib/protocols/jrmi.c]: will do

File: src/lib/protocols/jrmi.c (hunk 3 above)
Comment (dandare100) [reply src/lib/protocols/jrmi.c]: 100 %

File: src/lib/protocols/jrmi.c (hunk 4 above)
Comment (dandare100) [reply src/lib/protocols/jrmi.c]: Understood

File: src/lib/protocols/jrmi.c (hunk 6 above)
Comment (dandare100) [reply src/lib/protocols/jrmi.c]: Understood

**Comment by dandare100** [comment]
@IvanNardi 

I am not sure of the correct procedure to update tests as requested.
I have added a pcap and checked in the jrmi.pcap.out file.

I get an error on step 7 

```
Be sure to have nBPF support, cloning PF_RING in the same directory where you cloned nDPI: git clone https://github.com/ntop/PF_RING/ && cd PF_RING/userland/nbpf && ./configure && make
```
```
gcc -Wall -fPIC -O2 -I../lib -I../../kernel  -Wno-address-of-packed-member   -g nbpftest.c -o nbpftest  libnbpf.a `../lib/pfring_config --libs` -lpthread -lm 
/bin/sh: 1: ../lib/pfring_config: not found
```

**Comment by IvanNardi** [comment]
> ./lib/pfring_config: not found

You can ignore this error, because `linbbpf.a` has been created. I'll update the documentation to make that clear

In your specific case, at point 10, you can run `NDPI_FORCE_UPDATING_UTESTS_RESULTS=1 ./tests/do.sh` to update all the tests: "git add" the results

**Comment by dandare100** [comment]
Hello there @IvanNardi 
I have run the tests and there is no change to the original tests/cfgs/default/result/jrmi.pcap.out. 
Anything else I need to do ? 
I assume I must not add all the test results

**Comment by IvanNardi** [comment]
> I assume I must not add all the test results

That's exactly what you need to do :)
Your code triggers a (slightly) change in (almost) all the tests: we need to update all the results

**Comment by dandare100** [comment]
Eish. lol, my bad. I didn't want to blat over everyone's work. Done.

**Review by IvanNardi — APPROVED** [review]
I pushed some tiny changes.
Thanks for this contribution and for your patience!

**Comment by dandare100** [comment]
You are very welcome and thank you too

**Pull request merged**